File: chromium_src/components/payments/content/payment_request.cc
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "components/payments/content/payment_request.h"

#include "brave/components/brave_rewards/common/constants.h"
#include "brave/components/payments/content/buildflags/buildflags.h"
#include "third_party/blink/public/mojom/payments/payment_request.mojom.h"

using payments::mojom::PaymentErrorReason;

#if BUILDFLAG(ENABLE_PAY_WITH_BAT)
namespace payments {

void PaymentRequest::OnError(PaymentErrorReason reason, std::string err) {
  client_->OnError(reason, err);
}

void PaymentRequest::Init(
    mojo::PendingRemote<mojom::PaymentRequestClient> client,
    std::vector<mojom::PaymentMethodDataPtr> method_data,
    mojom::PaymentDetailsPtr details,
    mojom::PaymentOptionsPtr options) {
  std::vector<mojom::PaymentMethodDataPtr>::iterator it = std::find_if(
      method_data.begin(), method_data.end(), [](const auto& method) {
        return method->supported_method == brave_rewards::kBatPaymentMethod;
      });

  // Only BAT payment method needs additional checks
  if (it == method_data.end()) {
    Init_ChromiumImpl(std::move(client), std::move(method_data),
                      std::move(details), std::move(options));
    return;
  }

  // For BAT payment method, every item should have an associated SKU token
  if (details->display_items.has_value()) {
    for (const mojom::PaymentItemPtr& display_item : *details->display_items) {
      if (!(display_item->sku.has_value())) {
        log_.Error(brave_rewards::errors::kInvalidData);
        TerminateConnection();
        return;
      }
    }
  } else {
    log_.Error(brave_rewards::errors::kInvalidData);
    TerminateConnection();
    return;
  }

  Init_ChromiumImpl(std::move(client), std::move(method_data),
                    std::move(details), std::move(options));
}

}  // namespace payments

#define Init Init_ChromiumImpl
#endif
#include "../../../../../components/payments/content/payment_request.cc"

#if BUILDFLAG(ENABLE_PAY_WITH_BAT)
#undef Init
#endif

File: test/data/rewards-data/payments/payment_request.js
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

const supportedInstruments = [{
    supportedMethods: 'bat',
}];

const details = {
  id: '305DEF20-3F33-4D10-B901-8AFF34EBA750',
  total: {label: 'Total', amount: {currency: 'BAT', value: '0.25'}},
  displayItems: [
    {
      label: 'Brave T-Shirt',
      amount: {currency: 'BAT', value: '0.25'},
      sku: 'test-sku'
    },
  ],
};

const details_without_display_items = {
  id: '305DEF20-3F33-4D10-B901-8AFF34EBA750',
  total: {label: 'Total', amount: {currency: 'BAT', value: '0.25'}},
};

const details_without_sku_tokens = {
  id: '305DEF20-3F33-4D10-B901-8AFF34EBA750',
  total: {label: 'Total', amount: {currency: 'BAT', value: '0.25'}},
  displayItems: [
    {
      label: 'Brave T-Shirt',
      amount: {currency: 'BAT', value: '0.25'}
    },
  ],
};

var basic_request;
var request_without_display_items;
var request_without_sku_tokens;

function load() {
  basic_request = new PaymentRequest(supportedInstruments, details);
  request_without_display_items = new PaymentRequest(supportedInstruments, details_without_display_items);
  request_without_sku_tokens = new PaymentRequest(supportedInstruments, details_without_sku_tokens);
}

function print(msg) {
  document.getElementById('result').innerHTML = msg;
}

function run(testFunction) {
  try {
    testFunction().then(print).catch(print);
  } catch (error) {
    print("false");
  }
}

function batPaymentMethodSupported() {
  run(() => {
    return basic_request.canMakePayment();
  });
}

function paymentRequestWithoutDisplayItems() {
  run(() => {
    return request_without_display_items.canMakePayment();
  });
}

function paymentRequestWithoutSkuTokens() {
  run(() => {
    return request_without_sku_tokens.canMakePayment();
  });
}

